File: sections/26_tree_dp_basics/problems/b_road_pairing/solution.py
import sys


def main() -> None:
    data = list(map(int, sys.stdin.buffer.read().split()))
    if not data:
        return
    it = iter(data)
    n = next(it)
    adj = [[] for _ in range(n)]
    for _ in range(n - 1):
        u = next(it) - 1
        v = next(it) - 1
        adj[u].append(v)
        adj[v].append(u)
    parent = [-1] * n
    parent[0] = 0
    order = [0]
    for u in order:
        for v in adj[u]:
            if parent[v] == -1:
                parent[v] = u
                order.append(v)
    free = [0] * n
    used = [0] * n
    for u in reversed(order):
        base = 0
        children = []
        for v in adj[u]:
            if parent[v] == u:
                children.append(v)
                base += max(free[v], used[v])
        free[u] = base
        best = base
        for v in children:
            cand = base - max(free[v], used[v]) + free[v] + 1
            if cand > best:
                best = cand
        used[u] = best
    print(max(free[0], used[0]))


if __name__ == "__main__":
    main()

File: sections/26_tree_dp_basics/problems/b_road_pairing/solution.cpp
#include <bits/stdc++.h>
using namespace std;

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int n;
    if (!(cin >> n)) return 0;
    vector<vector<int>> adj(n);
    for (int i = 0; i < n - 1; i++) {
        int u, v;
        cin >> u >> v;
        --u; --v;
        adj[u].push_back(v);
        adj[v].push_back(u);
    }
    vector<int> parent(n, -1), order;
    parent[0] = 0;
    order.push_back(0);
    for (int i = 0; i < (int)order.size(); i++) {
        int u = order[i];
        for (int v : adj[u]) if (parent[v] == -1) {
            parent[v] = u;
            order.push_back(v);
        }
    }
    vector<int> free(n, 0), used(n, 0);
    for (int idx = n - 1; idx >= 0; idx--) {
        int u = order[idx];
        int base = 0;
        vector<int> children;
        for (int v : adj[u]) if (parent[v] == u) {
            children.push_back(v);
            base += max(free[v], used[v]);
        }
        free[u] = base;
        int best = base;
        for (int v : children) {
            best = max(best, base - max(free[v], used[v]) + free[v] + 1);
        }
        used[u] = best;
    }
    cout << max(free[0], used[0]) << '\n';
    return 0;
}

